### ISSUE TITLE
Question: Are the worker hooks thread safe?

### ISSUE THREAD

**Issue by Kangaroux** [benoitc/gunicorn#2490]
I am using a `.py` file for the config and there are some worker hooks. I want to track some information across the hooks and from the docs it's not clear to me if I need to use a mutex or not.

**Comment by benoitc** [comment]
worker's and server's hooks happen in only one thread context either the worker or the arbiter context. 

In async workers the `{pre,post}_request` hooks may not be threadsafe depending on what you're doing: they are executed in such case in the request context, can be a thread or a greenlet or anything else.

**Comment by Kangaroux** [comment]
Ok great, thanks!

**Comment by georgipopovhs** [comment]
For future record if anyone is looking at this issue, we have hit an issues where accessing `server.WORKERS.values()` from a hook (pre_fork) raised an exception `RuntimeError: dictionary changed size during iteration` (a solution could be to use a function running entirely in C and cannot be interrupted by releasing the GIL, such as `snapshot = server.WORKERS.copy()`).

The reason is that gunicorn handles signals such as SIGHUP (reloading gunicorn config and recycling workers), and the signal handler changes the `WORKERS` dict. Additionally, these handlers may also call hooks such as `pre_fork`.

In summary, even though it's a single thread, some hooks may be interrupted by a signal at any point, and the signal handler may enter the same hook or change state, which should be handled with care.
